In [2]:
import sys 
print(f"Python: {sys.version}")

import pandas as pd

Python: 3.8.0 (default, Nov  6 2019, 15:49:01) 
[Clang 4.0.1 (tags/RELEASE_401/final)]


In [3]:
df_bikes = pd.read_csv("./data/bike_rentals.csv")
df_bikes.head()

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
0        1  2011-01-01     1.0  0.0   1.0      0.0      6.0         0.0   
1        2  2011-01-02     1.0  0.0   1.0      0.0      0.0         0.0   
2        3  2011-01-03     1.0  0.0   1.0      0.0      1.0         1.0   
3        4  2011-01-04     1.0  0.0   1.0      0.0      2.0         1.0   
4        5  2011-01-05     1.0  0.0   1.0      0.0      3.0         1.0   

   weathersit      temp     atemp       hum  windspeed  casual  registered  \
0           2  0.344167  0.363625  0.805833   0.160446     331         654   
1           2  0.363478  0.353739  0.696087   0.248539     131         670   
2           1  0.196364  0.189405  0.437273   0.248309     120        1229   
3           1  0.200000  0.212122  0.590435   0.160296     108        1454   
4           1  0.226957  0.229270  0.436957   0.186900      82        1518   

    cnt  
0   985  
1   801  
2  1349  
3  1562  
4  1600

In [4]:
# Display numerical statistics
df_bikes.describe()

instant      season          yr        mnth     holiday     weekday  \
count  731.000000  731.000000  730.000000  730.000000  731.000000  731.000000   
mean   366.000000    2.496580    0.500000    6.512329    0.028728    2.997264   
std    211.165812    1.110807    0.500343    3.448303    0.167155    2.004787   
min      1.000000    1.000000    0.000000    1.000000    0.000000    0.000000   
25%    183.500000    2.000000    0.000000    4.000000    0.000000    1.000000   
50%    366.000000    3.000000    0.500000    7.000000    0.000000    3.000000   
75%    548.500000    3.000000    1.000000    9.750000    0.000000    5.000000   
max    731.000000    4.000000    1.000000   12.000000    1.000000    6.000000   

       workingday  weathersit        temp       atemp         hum   windspeed  \
count  731.000000  731.000000  730.000000  730.000000  728.000000  726.000000   
mean     0.682627    1.395349    0.495587    0.474512    0.627987    0.190476   
std      0.465773    0.544894    0.183094    0.163017    0.142331    0.077725   
min      0.000000    1.000000    0.059130    0.079070    0.000000    0.022392   
25%      0.000000    1.000000    0.336875    0.337794    0.521562    0.134494   
50%      1.000000    1.000000    0.499166    0.487364    0.627083    0.180971   
75%      1.000000    2.000000    0.655625    0.608916    0.730104    0.233218   
max      1.000000    3.000000    0.861667    0.840896    0.972500    0.507463   

            casual   registered          cnt  
count   731.000000   731.000000   731.000000  
mean    848.176471  3656.172367  4504.348837  
std     686.622488  1560.256377  1937.211452  
min       2.000000    20.000000    22.000000  
25%     315.500000  2497.000000  3152.000000  
50%     713.000000  3662.000000  4548.000000  
75%    1096.000000  4776.500000  5956.000000  
max    3410.000000  6946.000000  8714.000000

Comparing the mean and the median (50%) gives an indication of **skewness**. In the above results, mean and median are close to one another, so the data is roughly symmetrical.

In [5]:
# Display general information about the columns and rows
df_bikes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    float64
 3   yr          730 non-null    float64
 4   mnth        730 non-null    float64
 5   holiday     731 non-null    float64
 6   weekday     731 non-null    float64
 7   workingday  731 non-null    float64
 8   weathersit  731 non-null    int64  
 9   temp        730 non-null    float64
 10  atemp       730 non-null    float64
 11  hum         728 non-null    float64
 12  windspeed   726 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(10), int64(5), object(1)
memory usage: 91.5+ KB


*.info()* gives the number of rows, number of columns, column types, and non-nuşş values. Since the number of non-null values differs between columns, null values must be present in the dataset.
- If null values are not corrected, unexpected errors may arise down the road. 

In [8]:
# Display the total number of null values
df_bikes.isna().sum().sum() # First sum sums the null values in each col

12

In [13]:
df_bikes[df_bikes.isna().any(axis=1)]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
56        57  2011-02-26     1.0  0.0   2.0      0.0      6.0         0.0   
81        82  2011-03-23     2.0  0.0   3.0      0.0      3.0         1.0   
128      129  2011-05-09     2.0  0.0   5.0      0.0      1.0         1.0   
129      130  2011-05-10     2.0  0.0   5.0      0.0      2.0         1.0   
213      214  2011-08-02     3.0  0.0   8.0      0.0      2.0         1.0   
298      299  2011-10-26     4.0  0.0  10.0      0.0      3.0         1.0   
388      389  2012-01-24     1.0  1.0   1.0      0.0      2.0         1.0   
528      529  2012-06-12     2.0  1.0   6.0      0.0      2.0         1.0   
701      702  2012-12-02     4.0  1.0  12.0      0.0      0.0         0.0   
730      731  2012-12-31     1.0  NaN   NaN      0.0      1.0         0.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
56            1  0.282500  0.282192  0.537917        NaN     424        1545   
81            2  0.346957  0.337939  0.839565        NaN     203        1918   
128           1  0.532500  0.525246  0.588750        NaN     664        3698   
129           1  0.532500  0.522721       NaN   0.115671     694        4109   
213           1  0.783333  0.707071       NaN   0.205850     801        4044   
298           2  0.484167  0.472846  0.720417        NaN     404        3490   
388           1  0.342500  0.349108       NaN   0.123767     439        3900   
528           2  0.653333  0.597875  0.833333        NaN     477        4495   
701           2       NaN       NaN  0.823333   0.124379     892        3757   
730           2  0.215833  0.223487  0.577500   0.154846     439        2290   

      cnt  
56   1969  
81   2121  
128  4362  
129  4803  
213  4845  
298  3894  
388  4339  
528  4972  
701  4649  
730  2729

**df_bikes[conditional]** is a subset of *df_bikes* that meets the condition in brackets.
- In pandas, rows are **axis 0** and columns are **axis 1**
- **(axis=1)** specifies values in the columns.

`df_bikes.isna().any` gathers any and all null values.

Correcting null values depends on the column and dataset. Some useful strategies are:
1. **Replacing with the median/mean**
    - The idea here is to replace null values with the average column value.
    - To make changes permanent set `inplace=True` in `fillna()` method    

In [18]:
df_bikes.windspeed.fillna((df_bikes.windspeed.median()), inplace=True)
df_bikes.iloc[[56, 81, 128]] # rows 56, 81 and 128 were null for windpeed

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
56        57  2011-02-26     1.0  0.0   2.0      0.0      6.0         0.0   
81        82  2011-03-23     2.0  0.0   3.0      0.0      3.0         1.0   
128      129  2011-05-09     2.0  0.0   5.0      0.0      1.0         1.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
56            1  0.282500  0.282192  0.537917   0.180971     424        1545   
81            2  0.346957  0.337939  0.839565   0.180971     203        1918   
128           1  0.532500  0.525246  0.588750   0.180971     664        3698   

      cnt  
56   1969  
81   2121  
128  4362

*Tip: `.iloc` uses single brackets for one index as follows: `df_bikes.iloc[56]`. **df_bikes** also accepts list inside brackets to allow multiple indices. Multiple indices require double brackets: `df_bikes.iloc[[56, 81]]]`*

2. **Groupby with the median/mean**
    
    A groupby organizes rows by shared values.
    - These values are coming from different rows, so we need a way to combine, or aggregate, the values. Choices may include `.sum()`, `.count()`, `.mean()` and `.median()`.

    Lets correct the null values in the **hum** column, short for **humidity**, we can take the median humidity by season.

In [22]:
df_bikes.groupby(['season']).median() # Produced column values are medians

season
1.0    0.543750
2.0    0.646667
3.0    0.635833
4.0    0.661042
Name: hum, dtype: float64

In [20]:
df_bikes["hum"] = df_bikes["hum"].fillna(df_bikes.groupby("season")["hum"].transform('median'))

In [21]:
df_bikes.iloc[[129, 213, 388]]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
129      130  2011-05-10     2.0  0.0   5.0      0.0      2.0         1.0   
213      214  2011-08-02     3.0  0.0   8.0      0.0      2.0         1.0   
388      389  2012-01-24     1.0  1.0   1.0      0.0      2.0         1.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
129           1  0.532500  0.522721  0.646667   0.115671     694        4109   
213           1  0.783333  0.707071  0.635833   0.205850     801        4044   
388           1  0.342500  0.349108  0.543750   0.123767     439        3900   

      cnt  
129  4803  
213  4845  
388  4339

3. **Obtaining the median/mean from specific rows**
    
    In some cases, it may be advantageous to replace null values with data from specific rows.

    When correcting *temperature*, aside from consulting historical records, taking the mean temperature of the day before and the day after should give a good estimate.

In [24]:
df_bikes[df_bikes.temp.isna()]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
701      702  2012-12-02     4.0  1.0  12.0      0.0      0.0         0.0   

     weathersit  temp  atemp       hum  windspeed  casual  registered   cnt  
701           2   NaN    NaN  0.823333   0.124379     892        3757  4649

In [25]:
mean_temp = (df_bikes.iloc[700]["temp"] + df_bikes.iloc[700]["temp"])/2
mean_atemp = (df_bikes.iloc[700]["atemp"] + df_bikes.iloc[700]["atemp"])/2

df_bikes.temp.fillna((mean_temp), inplace=True)
df_bikes.atemp.fillna((mean_atemp), inplace=True)

df_bikes.iloc[701]

instant              702
dteday        2012-12-02
season               4.0
yr                   1.0
mnth                12.0
holiday              0.0
weekday              0.0
workingday           0.0
weathersit             2
temp            0.298333
atemp           0.316904
hum             0.823333
windspeed       0.124379
casual               892
registered          3757
cnt                 4649
Name: 701, dtype: object

In [26]:
df_bikes.iloc[[701]]

instant      dteday  season   yr  mnth  holiday  weekday  workingday  \
701      702  2012-12-02     4.0  1.0  12.0      0.0      0.0         0.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
701           2  0.298333  0.316904  0.823333   0.124379     892        3757   

      cnt  
701  4649

In [27]:
df_bikes[df_bikes.isna().any(axis=1)]

instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
730      731  2012-12-31     1.0 NaN   NaN      0.0      1.0         0.0   

     weathersit      temp     atemp     hum  windspeed  casual  registered  \
730           2  0.215833  0.223487  0.5775   0.154846     439        2290   

      cnt  
730  2729

4. Extrapolate Dates

    In order to correct null values involves dates, we have to first make sure that real dates are provided in the dataset. If so, date values may be extrapolated.

    `df_bikes["dteday"]` is a date column; however the type of the column is an object, commonly represented as a string. It may be converted to a **datetime**.

In [28]:
import datetime as dt

df_bikes["dteday"] = pd.to_datetime(df_bikes.dteday, infer_datetime_format=True)

df_bikes["mnth"] = df_bikes["dteday"].dt.month
df_bikes.tail()

instant     dteday  season   yr  mnth  holiday  weekday  workingday  \
726      727 2012-12-27     1.0  1.0    12      0.0      4.0         1.0   
727      728 2012-12-28     1.0  1.0    12      0.0      5.0         1.0   
728      729 2012-12-29     1.0  1.0    12      0.0      6.0         0.0   
729      730 2012-12-30     1.0  1.0    12      0.0      0.0         0.0   
730      731 2012-12-31     1.0  NaN    12      0.0      1.0         0.0   

     weathersit      temp     atemp       hum  windspeed  casual  registered  \
726           2  0.254167  0.226642  0.652917   0.350133     247        1867   
727           2  0.253333  0.255046  0.590000   0.155471     644        2451   
728           2  0.253333  0.242400  0.752917   0.124383     159        1182   
729           1  0.255833  0.231700  0.483333   0.350754     364        1432   
730           2  0.215833  0.223487  0.577500   0.154846     439        2290   

      cnt  
726  2114  
727  3095  
728  1341  
729  1796  
730  2729

As you can see, the month values are all correct, but the year value needs to be changed. Year column **yr** is normalized, meaning it's converted to values between 0 and 1. 
- Normalized data is often more efficient because machine learning weights do not have to adjust for different ranges.

In [29]:
df_bikes.loc[730, 'yr'] = 1.0

### CLEANING STEP 2: Deleting Non-numerical Columns

For machine learning, all data columns should be numerical. According to `.info()` method, the only column that is not numerical is `df_bikes["dteday"]`. Furthermore, it's redundant since all date information exists in other columns.

Thus, we should delete this columns as follows:

In [32]:
df_bikes = df_bikes.drop("dteday", axis=1)

## Predicting Regression

In the bike rentals dataset, `df_bikes["cnt"]` is the number of bike rentals in a given day. Predicting this column would be of great use to a bike rental company.

According to the dataset, `df_bikes["cnt"]` is the sum of `df_bikes["casual"]` and `df_bikes["registered"]`. Thus, it makes no sense to include these input columns that would be unknown in reality.

All current columns may be used to predict bike count except for abovementioned columns.

In [33]:
df_bikes = df_bikes.drop(["casual", "registered"], axis=1)

Now the dataset is cleaned and ready for machine learning task. It is best to save this form of dataset to file before going any further.

Afterwards, train and test sets will be created to be fed to ML pipeline.

In [34]:
# index = False parameter prevents an additional col from being created by the index
df_bikes.to_csv("bike_rentals_cleaned_170222.csv", index=False)

## Linear Regression

First model to be inspected is **LinearRegression** model. Linear Regression takes input columns as polynomial variables, and multiplies the values with coefficients (weights) to predict the target column. **Gradient Descent** works under the hood to minimize the error. The predictions of linear regression could be any real number.

In [36]:
from sklearn.model_selection import train_test_split
import warnings

warnings.filterwarnings('ignore')

X = df_bikes.iloc[:, :-1] # first : -> all the rows. second : -> from 1st col to the end without including it
y = df_bikes.iloc[:, -1]  # first : -> all the rows. second : -> only the last column

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=2)

In [39]:
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()

lin_reg.fit(X_train, y_train)

y_pred = lin_reg.predict(X_test)

Now compare the predictions with the test set. Scoring the model requires a basis of comparison. The standard for linear regression is the **root mean squared error (RMSE)**.
- The RMSE requires two pieces: *mean_squared_error*, the sum of the squares of differences between predicted and actual values, and *square root* to keep the units the same.

In [41]:
from sklearn.metrics import mean_squared_error
import numpy as np

mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)

print("RMSE: %0.2f" % rmse)

RMSE: 898.10


It is hard to know whether an error of **898** is good or bad without knowing the expected range of rentals per day.

RMSE has the same unit as dependent variable.

One important question to ask is *What is a good RMSE value?*. The short answer: *It depends on the range of the dataset*. The lower the RMSE, the better a given model is able to fit a dataset.

One way to gain a better understanding of whether a certain RMSE value is **good** is to normalize it using the formula:

`NormalizedRMSE = RMSE / (max - min)`

This produces a value between 0 and 1, where values closer to 0 represent better fitting models.

In [42]:
df_bikes.cnt.describe()

count     731.000000
mean     4504.348837
std      1937.211452
min        22.000000
25%      3152.000000
50%      4548.000000
75%      5956.000000
max      8714.000000
Name: cnt, dtype: float64

In [45]:
NormalizedRMSE = 898.10 / (8714 - 22)
NormalizedRMSE

0.10332489645651174

## XGBoost Regressor

In [47]:
from xgboost import XGBRegressor

xg_reg = XGBRegressor()
xg_reg.fit(X_train, y_train)

y_pred_xg = xg_reg.predict(X_test)

rmse_xg = np.sqrt(mean_squared_error(y_test, y_pred_xg))
print("RMSE: %0.2f" % rmse_xg)

RMSE: 691.98


In [48]:
NormalizedRMSE_XGB = 691.98 / (8714 - 22)
NormalizedRMSE_XGB

0.0796111366774045

XGBRegressor performs substantially better!

## Classification



In [50]:
# Column headings are the entires of first row, so load dataset with header=None
df_census = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', header=None)
df_census.head()

0                  1       2           3   4                    5   \
0  39          State-gov   77516   Bachelors  13        Never-married   
1  50   Self-emp-not-inc   83311   Bachelors  13   Married-civ-spouse   
2  38            Private  215646     HS-grad   9             Divorced   
3  53            Private  234721        11th   7   Married-civ-spouse   
4  28            Private  338409   Bachelors  13   Married-civ-spouse   

                   6               7       8        9     10  11  12  \
0        Adm-clerical   Not-in-family   White     Male  2174   0  40   
1     Exec-managerial         Husband   White     Male     0   0  13   
2   Handlers-cleaners   Not-in-family   White     Male     0   0  40   
3   Handlers-cleaners         Husband   Black     Male     0   0  40   
4      Prof-specialty            Wife   Black   Female     0   0  40   

               13      14  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3   United-States   <=50K  
4            Cuba   <=50K

In [51]:
# More info https://archive.ics.uci.edu/ml/datasets/Census+Income
df_census.columns=['age', 'workclass', 'fnlwgt', 'education', 'education-num', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week', 'native-country', 'income']
df_census.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [52]:
df_census.to_csv("census_income.csv", index=False)

### Step 2: Clean up Null Values

We can verify that there is no null values in the dataset.

In [53]:
df_census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [55]:
df_census.isna().sum().sum()

0

### Step 3: Drop Columns that have no information value

A quick review of the *df_census* data reveals that **education_num** column is a numerical conversion of **education** column. Since the information is the same we may delete the non-numeric one.

In [56]:
df_census = df_census.drop(["education"], axis=1)

### Step 4: Clean up Non-numerical Columns

All columns of the **dtype** object must be transformed into numerical columns. Pandas `get_dummies` method takes the non-numerical unique values of every column and converts them into their own column, with **1** indicating presence and **0** indicating absence. 

In [57]:
df_census = pd.get_dummies(df_census)
df_census.head()

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   workclass_ ?  workclass_ Federal-gov  workclass_ Local-gov  \
0             0                       0                     0   
1             0                       0                     0   
2             0                       0                     0   
3             0                       0                     0   
4             0                       0                     0   

   workclass_ Never-worked  ...  native-country_ Scotland  \
0                        0  ...                         0   
1                        0  ...                         0   
2                        0  ...                         0   
3                        0  ...                         0   
4                        0  ...                         0   

   native-country_ South  native-country_ Taiwan  native-country_ Thailand  \
0                      0                       0                         0   
1                      0                       0                         0   
2                      0                       0                         0   
3                      0                       0                         0   
4                      0                       0                         0   

   native-country_ Trinadad&Tobago  native-country_ United-States  \
0                                0                              1   
1                                0                              1   
2                                0                              1   
3                                0                              1   
4                                0                              0   

   native-country_ Vietnam  native-country_ Yugoslavia  income_ <=50K  \
0                        0                           0              1   
1                        0                           0              1   
2                        0                           0              1   
3                        0                           0              1   
4                        0                           0              1   

   income_ >50K  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 94 columns]

The target column is whether or not someone makes 50K. After `pd.get_dummies`, two columns **income_ <=50K** and **income_ >50K** are used to determine whether someone makes 50K. Since either column will work, we may delete one of them.

In [59]:
df_census = df_census.drop(["income_ <=50K"], axis=1)

### Split the data into X (predictor) and y (target) columns

In [60]:
X = df_census.iloc[:, :-1]
y = df_census.iloc[:, -1]

X.shape

(32561, 92)

### Apply Classification with Logistic Regression

Mathematically, logistic regression works in a manner similar to linear regression. For each column, logistic regression finds an appropriate weight that maximizes the model accuracy. The primary difference is that instead of summing each term, logistic regression uses the **sigmoid function**.

Let's use cross-validation on logistic regression to predict whether someone makes over 50K.

In [61]:
from sklearn.model_selection import cross_val_score

def cross_val(classifier, num_splits=10):
    model = classifier
    scores = cross_val_score(model, X, y, cv=num_splits)

    print(f"Accuracy: {np.round(scores, 2)}")
    print(f"Accuracy mean: {scores.mean()}")

In [62]:
from sklearn.linear_model import LogisticRegression
cross_val(LogisticRegression())

Accuracy: [0.8  0.8  0.79 0.8  0.79 0.81 0.79 0.79 0.8  0.8 ]
Accuracy mean: 0.7968121864153489


### Try with XGBoost Classifier

In [63]:
from xgboost import XGBClassifier
cross_val(XGBClassifier(n_estimators=5))

[00:36:36] WARNING: ../src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[00:36:37] WARNING: ../src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[00:36:37] WARNING: ../src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[00:36:38] WARNING: ../src/learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old be